File: go/heap.go
// LombokAlgoritma — binary min-heap used by the graph algorithms and Huffman coding
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

package lombokalgoritma

// minHeap is an array-backed binary min-heap ordered by less. Every caller uses a total
// lexicographic key (SPEC §9.0), so the pop order does not depend on the heap implementation.
type minHeap[T any] struct {
	items []T
	less  func(a, b T) bool
}

func newMinHeap[T any](less func(a, b T) bool) *minHeap[T] {
	return &minHeap[T]{less: less}
}

func (h *minHeap[T]) len() int { return len(h.items) }

func (h *minHeap[T]) push(x T) {
	a := append(h.items, x)
	i := len(a) - 1
	for i > 0 {
		p := (i - 1) >> 1
		if !h.less(a[i], a[p]) {
			break
		}
		a[i], a[p] = a[p], a[i]
		i = p
	}
	h.items = a
}

// pop removes and returns the minimum; ok is false when the heap is empty.
func (h *minHeap[T]) pop() (top T, ok bool) {
	a := h.items
	if len(a) == 0 {
		return top, false
	}
	top = a[0]
	last := len(a) - 1
	a[0] = a[last]
	a = a[:last]
	i := 0
	for {
		l := 2*i + 1
		r := l + 1
		m := i
		if l < len(a) && h.less(a[l], a[m]) {
			m = l
		}
		if r < len(a) && h.less(a[r], a[m]) {
			m = r
		}
		if m == i {
			break
		}
		a[i], a[m] = a[m], a[i]
		i = m
	}
	h.items = a
	return top, true
}

// tupleLess is lexicographic < on equal-length float tuples.
func tupleLess(a, b []float64) bool {
	for i := range a {
		if a[i] != b[i] {
			return a[i] < b[i]
		}
	}
	return false
}

File: rust/lombokalgoritma/src/core/heap.rs
// LombokAlgoritma — binary min-heap
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
//! Array-backed binary min-heap with a caller-supplied strict `less` predicate.
//!
//! SPEC §9.0 requires every graph heap to use a *total* lexicographic key (e.g. `[distance, node]`)
//! so the pop order — and therefore every output — does not depend on the heap implementation.
use alloc::vec::Vec;

/// Binary min-heap ordered by `less`.
pub struct MinHeap<T, F: Fn(&T, &T) -> bool> {
    items: Vec<T>,
    less: F,
}

impl<T, F: Fn(&T, &T) -> bool> MinHeap<T, F> {
    /// Empty heap ordered by the strict predicate `less`.
    pub fn new(less: F) -> Self {
        MinHeap {
            items: Vec::new(),
            less,
        }
    }

    /// Number of items.
    pub fn len(&self) -> usize {
        self.items.len()
    }

    /// `true` when the heap holds no items.
    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }

    /// Insert `item` (sift-up).
    pub fn push(&mut self, item: T) {
        let a = &mut self.items;
        a.push(item);
        let mut i = a.len() - 1;
        while i > 0 {
            let p = (i - 1) >> 1;
            if !(self.less)(&a[i], &a[p]) {
                break;
            }
            a.swap(i, p);
            i = p;
        }
    }

    /// Remove and return the minimum, or `None` when empty.
    pub fn pop(&mut self) -> Option<T> {
        let a = &mut self.items;
        let last = a.pop()?;
        if a.is_empty() {
            return Some(last);
        }
        let top = core::mem::replace(&mut a[0], last);
        let mut i = 0;
        loop {
            let l = 2 * i + 1;
            let r = l + 1;
            let mut m = i;
            if l < a.len() && (self.less)(&a[l], &a[m]) {
                m = l;
            }
            if r < a.len() && (self.less)(&a[r], &a[m]) {
                m = r;
            }
            if m == i {
                break;
            }
            a.swap(i, m);
            i = m;
        }
        Some(top)
    }
}

/// Lexicographic `<` on equal-length `f64` tuples, exactly like the TypeScript `tupleLess`
/// (the first differing component decides; `NaN` never compares less).
pub fn tuple_less(a: &[f64], b: &[f64]) -> bool {
    for (x, y) in a.iter().zip(b.iter()) {
        #[allow(clippy::float_cmp)]
        if x != y {
            return x < y;
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::vec;

    #[test]
    fn pops_in_order() {
        let mut h = MinHeap::new(|a: &[f64; 2], b: &[f64; 2]| tuple_less(a, b));
        assert!(h.is_empty());
        assert_eq!(h.pop(), None);
        for (i, d) in [5.0, 1.0, 3.0, 1.0, 9.0, 0.5, 3.0].iter().enumerate() {
            h.push([*d, i as f64]);
        }
        assert_eq!(h.len(), 7);
        let mut out = vec![];
        while let Some(x) = h.pop() {
            out.push(x);
        }
        assert_eq!(
            out,
            vec![
                [0.5, 5.0],
                [1.0, 1.0],
                [1.0, 3.0],
                [3.0, 2.0],
                [3.0, 6.0],
                [5.0, 0.0],
                [9.0, 4.0]
            ]
        );
        assert!(!tuple_less(&[1.0, 2.0], &[1.0, 2.0]));
        assert!(!tuple_less(&[f64::NAN], &[1.0]));
    }
}
